app/main.py

```diff
@@ -1,2 +1,204 @@
+from __future__ import annotations
+from dataclasses import dataclass
+from fractions import Fraction
+from typing import Hashable, Any, Iterator, Optional, Iterable
+
+
 class Dictionary:
-    pass
+    @dataclass
+    class Node:
+        key: Hashable
+        value: Any
+        key_hash: int
+
+    INITIAL_CAPACITY = 8
+    RESIZE_THRESHOLD = Fraction(2, 3)
+    CAPACITY_MULTIPLIER = 2
+
+    def __init__(self, capacity: int = INITIAL_CAPACITY) -> None:
+        self.capacity = capacity
+        self.size = 0
+        self.hash_table: list[Dictionary.Node | None] = [None] * self.capacity
+
+    def _calculate_index(self, key: Hashable) -> int:
+        index = hash(key) % self.capacity
+
+        while (
+            self.hash_table[index] is not None
+            and self.hash_table[index].key != key
+        ):
+            index = (index + 1) % self.capacity
+
+        return index
+
+    @property
+    def current_max_size(self) -> int:
+        return int(self.capacity * self.RESIZE_THRESHOLD)
+
+    def _resize(self) -> None:
+        old_hash_table = self.hash_table
+
+        self.__init__(self.capacity * self.CAPACITY_MULTIPLIER)
+
+        for node in old_hash_table:
+            if node is not None:
+                self.__setitem__(node.key, node.value)
+
+    def __setitem__(self, key: Hashable, value: Any) -> None:
+        index = self._calculate_index(key)
+        key_hash = hash(key)
+
+        if self.hash_table[index] is None:
+            if self.size + 1 >= self.current_max_size:
+                self._resize()
+                index = self._calculate_index(key)
+            self.size += 1
+
+        self.hash_table[index] = Dictionary.Node(key, value, key_hash)
+
+    def __getitem__(self, key: Hashable) -> Any:
+        index = self._calculate_index(key)
+
+        if self.hash_table[index] is None:
+            raise KeyError(f"Cannot find value for key: {key}")
+
+        return self.hash_table[index].value
+
+    def __delitem__(self, key: Hashable) -> None:
+        index = self._calculate_index(key)
+
+        if self.hash_table[index] is None:
+            raise KeyError(f"Cannot delete value for key: {key}")
+
+        self.hash_table[index] = None
+        self.size -= 1
```

Comment:
you need to recalculate remaining positions after deleting an element because indexes may be wrong for elements that had collisions